src/TTextEdit.cpp

```diff
@@ -983,37 +989,77 @@ void TTextEdit::mouseMoveEvent(QMouseEvent* event)
+        int column = 0;
+        // This is the calculated (estimated) cumulative width of all of the
+        // graphemes considered in the line so far:
         int currentX = 0;
-        for (const auto& character : mpBuffer->lineBuffer.at(lineNumber)) {
-            const uint unicode = getGraphemeBaseCharacter(character);
+        QString lineText = mpBuffer->lineBuffer.at(lineNumber);
+        QTextBoundaryFinder boundaryFinder(QTextBoundaryFinder::Grapheme, lineText);
+#if defined (QT_DEBUG)
+        if (debug_mousePositionInLine) {
+            qDebug().noquote().nospace() << "TTextEdit::convertMouseXToBufferX(" << mouseX << ", " << lineNumber << ") INFO - finding mouse click position in line:";
+        }
+#endif
+        for (int indexOfChar = 0, total = lineText.size(); indexOfChar < total; /*incrementing is done inside loop*/) {
+            int nextBoundary = boundaryFinder.toNextBoundary();
+            // Width in "normal" width equivalent of this grapheme:
+            int charWidth = 0;
+            const QString grapheme = lineText.mid(indexOfChar, nextBoundary - indexOfChar);
+#if defined (QT_DEBUG)
+            if (debug_mousePositionInLine) {
+                if (indexOfChar > 0) {
+                    // Include graphemes BEFORE the current last one considered:
+                    qDebug().noquote().nospace() << lineText.left(indexOfChar) << "»" << grapheme << "«";
+                } else {
+                    // First grapheme:
+                    qDebug().noquote().nospace() << "»" << grapheme << "«";
+                }
+            }
+#endif
+            const uint unicode = getGraphemeBaseCharacter(grapheme);
             if (unicode == '\t') {
-                characterWidth = 8;
+                charWidth = mTabStopwidth - (column % mTabStopwidth);
             } else {
-                characterWidth = getGraphemeWidth(unicode);
+                charWidth = getGraphemeWidth(unicode);
             }
-            currentX += characterWidth * mFontWidth;
-            characterIndex++;
+            column +=charWidth;
+            currentX = column * mFontWidth;
             if (currentX >= mouseX) {
                 if (mShowTimeStamps) {
-                    characterIndex -= 13;
+                    // The timestamp is (currently) 13 "normal width" characters
+                    // but that might not always be the case in some future I18n
+                    // situations:
+                    return std::max(0, indexOfChar - 13);
                 }
-                characterIndex = std::max(characterIndex, 0);
-                return characterIndex;
+                return std::max(0, indexOfChar);
             }
+            // Increment for loop:
+            indexOfChar = nextBoundary;
         }
     }
 
-    characterIndex = mouseX / mFontWidth;
-    if (mShowTimeStamps) {
-        characterIndex -= 13;
-    }
-    characterIndex = std::max(characterIndex, 0);
-    return characterIndex;
```

Comment:
Why was all of this deleted? This keeps the old behaviour of still returning a theoretical X position for when you're out of the buffer range. I don't know how the code will behave when it's suddenly getting `0` now, have you tested and verified all of this?

Reply:
If the `lineNumber` argument is out of range then there is **NO** useful value to return! The "BufferX" in the method name is not a spatial characteristic but an array index - note that the value I return is based on `indexOfChar` which is the first `QChar` in the particular grapheme under the mouse cursor on the relevant line {perhaps that should have been `indexOfQChar`} and it would not be possible to derive a meaningful value in the absence of any graphemes to draw and estimate their width from.

*Perhaps* that is not entirely fair: it *might* be possible to estimate the number of *average* width characters - given that that is how much space we now assign for a single, "normal" width grapheme - but to than take that number and to use it to index into a random line of text in the `TBuffer::lineBuffer` would be *nonsensical* even if it did not cause a seg. fault by trying to access beyond the end of that random line.

Thus in the situation when it is not possible to access the buffer either we could return `-1` as a sentinel value as I hypothesize or **we return `0` (as we do currently) and rely on the caller to check that `lineNumber` of the buffer will support the index value returned**...

Reply:
Ok but that's how the rest of the code works right now and I don't want to refactor it entirely and or introduce a behaviour change. It's one day until feature freeze, now is not the time

Reply:
I've taken a closer look at b3f37615ed41306393da55d1d1df7e061762f2da and what you are saying about "*that is how the rest of the code works right now*" seems to confirm to me that that is why the existing code *doesn't* work for non-BMP and combining diacriticals precisely because there is *not* a 1:1 relationship between how far across from the left margin the mouse X position is and the index into the `QChar`s that form the `QString` that is the textual data for each line of Mud text.  By 1:1 I mean that for a certain number of pixels from the left side the index of the `QChar` that is at that number of pixels is a fixed and invariant value. That is what the previous code assumed and it is only true for BMP characters without any diacriticals (and without any non-printing Unicode codepoints such as Variation Selectors, Joiners, etc.)

The code I am proposing is good enough provided that we assume that the text we present (barring tabs) can be done in a duo-spaced (each glyph takes up either one or two "spaces") manner. {As it happens I have prototyped code that does things in a slower and more involved manner that actually recorded the width each grapheme took when painted on screen and looked that up when it need to work out where the mouse movement put the cursor - which meant it could handle Qt using glyphs from different fonts as needed to display the text - which this code will not cope with too well if the individual characters deviate *too* much from the "average" width.}

Saying "it's is one day until feature freeze" is not helpful - either we want code that works for a non-English user base or we put forward a product that has known miss-features.

I am sorry if I did not point this out more strongly when the 4.0.0. release was recently discussed but as I said at the time I did not feel we had everything sorted yet and I pointed out the limits on my current attention to the project and the reason why. It was only when I started looking at your PR - that this one is hacking away at - that I realised that I hadn't sorted out the selection and highlighting code as completely as I had intended. OTOH This will, I believe, be good enough for Western and Chinese languages cases.

Reply:
> That is what the previous code assumed and it is only true for BMP characters without any diacriticals (and without any non-printing Unicode codepoints such as Variation Selectors, Joiners, etc.)

I'm sorry but I don't follow what are you talking about. We've ruled out emojis as not a 4.0 requirement, so what usecases are you looking to fix here?

My original PR fixes it for Chinese, all the other languages we know of work fine. As far as I know there are no Vietnamese MUDs so that is not a concern. Is the concern you're having practical or theoretical?